Match on variables instead of expressions (#877)
unicode-org/message-format-wg: spec/syntax.md

- The number of _keys_ on each _variant_ MUST be equal to the number of _selectors_.
- At least one _variant_ MUST exist whose _keys_ are all equal to the "catch-all" key `*`.
- Each _selector_ MUST be a _variable_ that
  directly or indirectly references a _declaration_ with an _annotation_.
- Each _variant_ MUST use a list of _keys_ that is unique from that
  of all other _variants_ in the _message_.
  _Literal_ _keys_ are compared by their contents, not their syntactical appearance.

```abnf
matcher         = match-statement s variant *([s] variant)

> match-statement = match 1*(s selector)

Reviewer: Ibid if my proposed change is accepted

```

> A _message_ with a _matcher_:
>
> ```
> .input {$count :number}
> .match $count
> one {{You have {$count} notification.}}
> *   {{You have {$count} notifications.}}
> ```

> A _message_ containing a _matcher_ formatted on a single line:
>
> ```
> .local $os = {:platform} .match $os windows {{Settings}} * {{Preferences}}
> ```

### Selector

> A **_<dfn>selector</dfn>_** is a _variable_ that ranks or excludes the

Reviewer: This is doesn't make sense?

```suggestion
A **_<dfn>selector</dfn>_** is a _variable_ whose resolved value contains an _expression_
that ranks or excludes the
```

Or perhaps instead of "contains" say "is a _variable_ annotated with an _expression_"?

Author: This could indeed be imprvoed, but I think including _expression_ here would be misleading, as the selection is ultimately done by the value coming from the function handling the annotation.
```suggestion
A **_<dfn>selector</dfn>_** is a _variable_ whose resolved value ranks or excludes the
```

Reviewer: > as the selection is ultimately done by the value coming from the function handling the annotation

We sometimes talk as if this is the case, but it's not true. The function doesn't emit a value (unless that value is sorted-and-filtered list of keys from `MatchSelectorsKeys(r,k)`). The variable here is the operand being used by the selector _plus_ the annotation configuring the selector. A function name like `:number` has two separate interfaces: one for formatting and one for selection.

I agree that my formulation is kind of klunky, but one of the ramifications of Option F is that there MUST be an annotating expression in a declaration.

Maybe:

```suggestion
A **_<dfn>selector</dfn>_** is a _variable_ that appears in a
previous _declaration_ and whose resolved value ranks or excludes the
```
